lib/ansible/modules/system/osx_defaults.py

```diff
@@ -15,413 +16,975 @@
 
-    # tools --------------------------------------------------------------- {{{
-    """ Converts value to given type """
+    # List of all text characters.
+    TEXT_CHARS = bytearray(set([7, 8, 9, 10, 12, 13, 27]) | set(range(0x20, 0x100)) - set([0x7f]))
 
-    def _convert_type(self, type, value):
+    def __new__(cls, data):
+        if isinstance(data, cls):
+            return data
 
-        if type == "string":
-            return str(value)
-        elif type in ["bool", "boolean"]:
-            if isinstance(value, (binary_type, text_type)):
-                value = value.lower()
-            if value in [True, 1, "true", "1", "yes"]:
-                return True
-            elif value in [False, 0, "false", "0", "no"]:
-                return False
-            raise OSXDefaultsException("Invalid boolean value: {0}".format(repr(value)))
-        elif type == "date":
+        # Try to convert unicode to ascii string.
+        if isinstance(data, text_type):
             try:
-                return datetime.datetime.strptime(value.split("+")[0].strip(), "%Y-%m-%d %H:%M:%S")
-            except ValueError:
-                raise OSXDefaultsException(
-                    "Invalid date value: {0}. Required format yyy-mm-dd hh:mm:ss.".format(repr(value))
-                )
-        elif type in ["int", "integer"]:
-            if not str(value).isdigit():
-                raise OSXDefaultsException("Invalid integer value: {0}".format(repr(value)))
-            return int(value)
-        elif type == "float":
-            try:
-                value = float(value)
-            except ValueError:
-                raise OSXDefaultsException("Invalid float value: {0}".format(repr(value)))
-            return value
-        elif type == "array":
-            if not isinstance(value, list):
-                raise OSXDefaultsException("Invalid value. Expected value to be an array")
-            return value
+                data = data.encode('ascii')
+            except UnicodeEncodeError:
+                pass
 
-        raise OSXDefaultsException('Type is not supported: {0}'.format(type))
+        # Check if data is a valid base64 string. Short strings are not
+        # considered as binary.
+        if isinstance(data, binary_type) and len(data) > 51 and not data.translate(None, cls.BASE64_CHARS):
+            try:
+                binary_data = b64decode(data)
+            except TypeError:
+                pass
+            else:
+                if cls.is_binary(binary_data):
+                    return super(Data, cls).__new__(cls, data)
 
-    """ Returns a normalized list of commandline arguments based on the "host" attribute """
+        raise ValueError('Unsupported data type.')
 
-    def _host_args(self):
-        if self.host is None:
-            return []
-        elif self.host == 'currentHost':
-            return ['-currentHost']
+    @classmethod
+    def is_binary(cls, data):
```

Comment:
This looks like a rather provisional check to me. Why just the first 512 characters/bytes? Why 30%?

Reply:
This part is completely inspired by this:
http://code.activestate.com/recipes/173220-test-if-a-file-or-string-is-text-or-binary/

The 512 chars are just an optimization because there's very little chance that binary data will not have a good proportion of non-text characters in the first 512 characters.

For the 30%, I just follow the reference. I can't justify this number.

Here also, if someone have a better/simpler solution I'm all open. I did a lot of researches on this and it's the best solution I found.

Reply:
Any chance on simply requiring that anywhere where some kind of binary data *might* be provided that *all* such data must be `base64`-encoded? Even if it's actual text? Can you identify if property can/should be binary data?

Reply:
My first motivation to work on this module was to add access to nested values and to be able to support dicts and lists. Lists and dicts can be nested. Supporting lists and dicts imply two things:
- It's no longer possible to easily specify the type of each items contains in dicts and lists.
- You need to support in some ways all the types supported by the plist because you need to compare existing lists/dicts to the wanted values and those lists/dicts can contains values of any types. That means supporting binary data and datetime.

My first idea to answer the first point was to use the YAML casting property to send values already in their own type. Unfortunately, it just partly works because, for example, values are sent to the host in JSON and JSON do not support binary data and datetime. So I'm sending them as string and deduce their type on arrival.

A nicer solution to this problem would be that Ansible will define a way to send binary data and datetime in JSON (for example by using a marker at the start of the string that tell the type) or by sending YAML instead of JSON.

To answer your question more directly: it's already required that all binary data should be `base64`-encoded. I can identify if a property should be binary only if it's already existing in the plist. For new values, there's no way to know. plist files are just dict/list for storing values. Each app treat plist as they want.

Reply:
Hm... How about then introducing a prefix for such settings at least for this module? So, if user wants to provide binary data to be stored in property, they'd have to say, e.g. 'B64:blah' (where blah would be base64-encoded data). Caveat being that if users wants to store string which starts with B64:, they'd have to do something along the lines of B64:QjY0Og== (QjY0Og== being base64-encoded B64:, I think you get the picture :)

Now, truth be told, not sure if this would be more acceptable for Ansible as project, but I'd see it as more consistent. Maybe second opinions on this could be useful too :)

Reply:
That's an idea. Another option for the prefix could be `!!binary`. It's [what is use in YAML](http://yaml.org/type/binary.html). So, if one day Ansible support sending binary data, people will just have to move the first quote after the `!!binary` (`"!!binary R0lGODlhDAA"` -> `!!binary "R0lGODlhDAA"`). Definitely not sure if it's a good idea!

But second opinions is definitely needed on this.